plugin-server/src/worker/plugins/loadPlugin.ts

```diff
@@ -70,22 +71,69 @@ export async function loadPlugin(server: Hub, pluginConfig: PluginConfig): Promi
+            if (
+                plugin.source_frontend &&
+                (plugin.transpiled_frontend === null || plugin.transpiled_frontend === undefined)
+            ) {
+                if (await getTranspilationLock(hub, plugin)) {
+                    const transpilationStartTimer = new Date()
+                    try {
+                        plugin.transpiled_frontend = transpileFrontend(plugin.source_frontend)
+                    } catch (error: any) {
+                        plugin.transpiled_frontend = "'error'"
+                        await processError(hub, pluginConfig, error)
+                        hub.statsd?.increment(`transpile_frontend.ERROR`, {
+                            plugin: plugin.name ?? '?',
+                            pluginId: `${plugin.id ?? '?'}`,
+                        })
+                    }
+                    hub.statsd?.timing(`transpile_frontend`, transpilationStartTimer, {
+                        plugin: plugin.name ?? '?',
+                        pluginId: `${plugin.id ?? '?'}`,
+                    })
+
+                    await hub.db.postgresQuery(
+                        'update posthog_plugin set transpiled_frontend = $1 where id = $2',
+                        [plugin.transpiled_frontend, plugin.id],
+                        'setPluginTranspiledFrontend'
+                    )
+                }
+            }
+
+            if (plugin.source) {
+                void pluginConfig.vm?.initialize!(plugin.source, pluginDigest(plugin))
+                return true
+            }
+
+            if (!plugin.source && !plugin.transpiled_frontend) {
+                pluginConfig.vm?.failInitialization!()
+                await processError(hub, pluginConfig, `Could not load source code for ${pluginDigest(plugin)}!`)
             }
-        } else if (plugin.plugin_type === 'source' && plugin.source) {
-            void pluginConfig.vm?.initialize!(plugin.source, pluginDigest(plugin))
-            return true
         } else {
             pluginConfig.vm?.failInitialization!()
             await processError(
-                server,
+                hub,
                 pluginConfig,
                 `Tried using undownloaded remote ${pluginDigest(plugin)}, which is not supported!`
             )
         }
     } catch (error) {
         pluginConfig.vm?.failInitialization!()
-        await processError(server, pluginConfig, error)
+        await processError(hub, pluginConfig, error)
     }
     return false
 }
+
+export async function getTranspilationLock(hub: Hub, plugin: Plugin): Promise<boolean> {
+    // Only update `transpiled_frontend` if it equals NULL at the time of the update.
+    const response = await hub.db.postgresQuery(
+        'UPDATE posthog_plugin SET transpiled_frontend = $1 ' +
+            'WHERE id = $2 AND transpiled_frontend IS NULL RETURNING transpiled_frontend',
+        ["'transpiling'", plugin.id],
+        'getPluginTranspiledFrontendLock'
```

Comment:
What happens when someone fetches this when `transpiled_frontend` is `"transpiling"`?